### PR TITLE
Feature: exponential height Fog

### PR DESCRIPTION
**Pull request by Bernatmago**
**Changes**
- Moved ambient light params to scene and added serialization
- Added exponential height fog + scene params and serialization
- Reorganized scene config menus a bit
- Changed batches so they work when we call draw consecutive times (previously it needed reset before each draw)

To easily do the fog i add the transparent meshes positions to the g buffer since the lighting pass has already been done. I do it this way because that's the easiest place to read all the fragment positions from (needed for fog). We need the real values to prevent this from happening when there are transparent meshes. We also prevent reserving extra memory doing it this way, in the same pass (adding transparent fragment possitions to g buffer) we can add any transparent fragment data that's needed for postprocessing to have easy access to all the values we might need.
image
image

Quick fog examples
image
image

### CODE REVIEW

**Review by monikamnc — APPROVED** [review]
Code looking good to me but it will be great at least one of our render boys check it.
@miquelmiro3 @baransrc 
Tested and working fine, I had a bug related to vector memory size that I leave in discord and I can't replicate it again uwu

One thing that I miss is that when I press play (play the game then stop) I lost the fog configuration because I imagine that it loads the serialized one so until I don't press save scene we always lose the changes? Or can be saved somewhere (?)

Thank you for your work and nice return to the render team boys ^-^

**Comment by Bernatmago** [comment]
> Code looking good to me but it will be great at least one of our render boys check it. @miquelmiro3 @baransrc Tested and working fine, I had a bug related to vector memory size that I leave in discord and I can't replicate it again uwu
> 
> One thing that I miss is that when I press play (play the game then stop) I lost the fog configuration because I imagine that it loads the serialized one so until I don't press save scene we always lose the changes? Or can be saved somewhere (?)
> 
> Thank you for your work and nice return to the render team boys ^-^

What you mention sounds like a bug bcs the idea was that the scene saved a temporary copy when pressing play and used it on stop ptress, we'll have to check it

File: Source/src/Batching/BatchManager.cpp [hunk 1]
```diff
@@ -192,9 +192,13 @@ void Hachiko::BatchManager::ShowDebugMenuForBatches(
 void Hachiko::BatchManager::DrawSingleBatch(GeometryBatch* geometry_batch, const Program* program, int segment) const 
 {
     // Update and bind batch
-    geometry_batch->UpdateCommands();
-    geometry_batch->UpdateBatch(segment);
-    geometry_batch->BindBatch(segment, program);
+    if (geometry_batch->dirty_draw_components)
```
Comment (miquelmiro3) [inline Source/src/Batching/BatchManager.cpp:195]: Why did you add this check?

**Review by miquelmiro3 — APPROVED** [review]
All looking good! Gj

File: Game/shaders/fragment_fog.glsl [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+uniform vec3 fog_color;
+uniform float fog_global_density;
+uniform float fog_height_falloff;
+
+layout (binding = 3) uniform sampler2D g_position;
+
+void main()
+{
+    vec3  fragment_position = (texture(g_position, texture_coords)).rgb;
+    // vec3 view_direction = normalize(camera.pos - fragment_position);
```
Comment (baransrc) [inline Game/shaders/fragment_fog.glsl:20]: Let's delete this one.

File: Source/src/modules/ModuleRender.cpp [hunk 3]
```diff
@@ -365,35 +366,64 @@ void Hachiko::ModuleRender::DrawDeferred(Scene* scene, ComponentCamera* camera,
...
+            glBindFramebuffer(GL_DRAW_FRAMEBUFFER, frame_buffer);
+        }
     }
-    
-    // Get the targets that has transparent materials. These targets will be
-    // rendered with regular forward rendering pass:
-    for (const RenderTarget& target : transparent_targets)
+
+    // ----------------------------- FOG -----------------------------
+
```
Comment (baransrc) [inline Source/src/modules/ModuleRender.cpp:403]: Let's put these fog pass in a single method and call it instead.

File: Source/src/modules/ModuleRender.cpp [hunk 4]
```diff
@@ -365,35 +366,64 @@ void Hachiko::ModuleRender::DrawDeferred(Scene* scene, ComponentCamera* camera,
...
+        program->Activate();
 
-    // Forward rendering pass for transparent game objects:
-    program = App->program->GetForwardProgram();
-    program->Activate();
+        EnableBlending();
+        
+        program->BindUniformFloat3("fog_color", scene->GetFogColor().ptr());
+        program->BindUniformFloat("fog_global_density", scene->GetFogGlobalDensity());
+        program->BindUniformFloat("fog_height_falloff", scene->GetFogHeightFalloff());
```
Comment (baransrc) [inline Source/src/modules/ModuleRender.cpp:417]: See Uniforms.h for these type of uniform names, across the engine, we are now trying to gather all uniform names in a global space and use those instead of these "magic strings". On Uniforms.h you can add a namespace inside that is called Fog, and add these uniform binding names there as constexprs.

File: Source/src/modules/ModuleSceneManager.cpp [hunk 5]
```diff
@@ -309,11 +309,14 @@ void Hachiko::ModuleSceneManager::OptionsMenu()
...
-    // Skybox
     ImGui::Separator();
-    main_scene->GetSkybox()->DrawImGui();
+    main_scene->AmbientLightOptionsMenu();
+    ImGui::Separator();
+    main_scene->FogOptionsMenu();
+    ImGui::Separator();
+    main_scene->SkyboxOptionsMenu();
+    
+    // App->navigation->DrawOptionsGui();
```
Comment (baransrc) [inline Source/src/modules/ModuleSceneManager.cpp:319]: Let's remove this comment.

File: Source/src/core/Scene.cpp [hunk 6]
```diff
@@ -281,6 +283,65 @@ void Hachiko::Scene::Load(const YAML::Node& node, bool meshes_only)
     loaded = true;
 }
 
+void Hachiko::Scene::LoadAmbientParams(const YAML::Node& node)
+{
+    if (!node[AMBIENT_LIGHT].IsDefined()) return;
```
Comment (baransrc) [inline Source/src/core/Scene.cpp:288]:
```suggestion
    if (!node[AMBIENT_LIGHT].IsDefined()) 
    {
        return;
    }
```

File: Source/src/core/Scene.cpp [hunk 7]
```diff
@@ -281,6 +283,65 @@ void Hachiko::Scene::Load(const YAML::Node& node, bool meshes_only)
...
+void Hachiko::Scene::SaveAmbientParams(YAML::Node& node)
+{
+    YAML::Node ambient_node = node[AMBIENT_LIGHT];
+    ambient_node[AMBIENT_LIGHT_INTENSITY] = ambient_light_intensity;
+    ambient_node[AMBIENT_LIGHT_COLOR] = ambient_light_color;
+}
+
+void Hachiko::Scene::LoadFogParams(const YAML::Node& node)
+{
+    if (!node[FOG].IsDefined()) return;
```
Comment (baransrc) [inline Source/src/core/Scene.cpp:303]:
```suggestion
    if (!node[FOG].IsDefined())
    {
        return;
    }
```

File: Source/src/core/Scene.h [hunk 8]
```diff
@@ -116,8 +115,46 @@ namespace Hachiko
...
+
+        const float* GetFogGlobalDensity() const
+        {
+            return &fog_global_density;
+        }
+
+        const float* GetFogHeightFalloff() const 
+        {
+            return &fog_height_falloff;
+        }
```
Comment (baransrc) [inline Source/src/core/Scene.h:146]: I would encapsulate fog related stuff inside a small struct/class that is called FogConfig and have scene own one to not convolute the scene class any further maybe. Then the API would be `Scene::fog_config.IsFogEnabled` for example.

File: Source/src/modules/ModuleProgram.cpp [hunk 9]
```diff
@@ -35,14 +35,16 @@ bool Hachiko::ModuleProgram::Init()
...
+    CreateFogProgram();
 
     if (!forward_program || !deferred_geometry_program || 
         !deferred_lighting_program || !skybox_program || 
         !diffuse_ibl_program || !prefiltered_ibl_program || 
         !environment_brdf_program || !stencil_program || 
         !ui_image_program || !ui_text_program || 
         !particle_program || !shadow_mapping_program ||
-        !gaussian_filtering_program)
+        !gaussian_filtering_program || !transparent_depth_program || !fog_program)
```
Comment (baransrc) [inline Source/src/modules/ModuleProgram.cpp:47]: I love how chad this if is.

**Review by baransrc — CHANGES_REQUESTED** [review]
Thanks for the solid piece of work!! Welcome back to dark side (renderboiis), there are some small change requests and some suggestions, after you address those, we're good to merge this bad boy. 💯

File: Source/src/Batching/BatchManager.cpp (hunk 1 above)
Comment (Bernatmago) [reply Source/src/Batching/BatchManager.cpp:195]: Because i call this method two times in a row and it was a waste of resources to repeat the process

File: Source/src/core/Scene.h (hunk 8 above)
Comment (Bernatmago) [reply Source/src/core/Scene.h:146]: Done <3

**Review by monikamnc — APPROVED** [review]
